File: ios/StoreLink/StoreLinkCore.swift
//
//  StorelinkCore.swift
//  Storelink
//
import Foundation
import UIKit

public class StorelinkCore {
    
    // Configuration for the SDK
    public struct Configuration {
        let refreshToken: String
        let logLevel: LogLevel?
        let onConfigurationSuccess: (([AnyHashable: Any]) -> Void)?
        let onStoreConnectionEvent: (([AnyHashable: Any]) -> Void)?
        let onInvoiceEvent: (([AnyHashable: Any]) -> Void)?
        let onCheckingStoreConnectionEvent: (([AnyHashable: Any]) -> Void)?
        let brandName: String?
        let logoUrl: String?
        let _devApiLocation: String?
        // Add a public initializer
        public init(
            refreshToken: String,
            logLevel: LogLevel? = nil,
            onConfigurationSuccess: (([AnyHashable: Any]) -> Void)? = nil,
            onStoreConnectionEvent: (([AnyHashable: Any]) -> Void)? = nil,
            onInvoiceEvent: (([AnyHashable: Any]) -> Void)? = nil,
            onCheckingStoreConnectionEvent: (([AnyHashable: Any]) -> Void)? = nil,
            brandName: String? = nil,
            logoUrl: String? = nil,
            _devApiLocation: String? = nil
        ) {
            self.refreshToken = refreshToken
            self.logLevel = logLevel
            self.onConfigurationSuccess = onConfigurationSuccess
            self.onStoreConnectionEvent = onStoreConnectionEvent
            self.onInvoiceEvent = onInvoiceEvent
            self.onCheckingStoreConnectionEvent = onCheckingStoreConnectionEvent
            self.brandName = brandName
            self.logoUrl = logoUrl
            self._devApiLocation = _devApiLocation
        }
    }
    
    // Possible errors for the SDK
    public enum SDKError: Error {
        case invalidRefreshToken
        case unknownError
        case nativeCommunicationError
    }
    
    public enum PresentationMethod {
        case presentModally(on: UIViewController)
        case pushOnNavigationStack(navigationController: UINavigationController)
        case embedInTab(tabBarController: UITabBarController, at: Int?)
    }
    
    
    // Define a handler that the SDK uses
    public class SDKHandler {
        
        private var config: Configuration
        private var notificationName = Notification.Name("cooklist_sdk_event")
        private var onConfigurationSuccess: (([AnyHashable: Any]) -> Void)?
        private var onStoreConnectionEvent: (([AnyHashable: Any]) -> Void)?
        private var onInvoiceEvent: (([AnyHashable: Any]) -> Void)?
        private var onCheckingStoreConnectionEvent: (([AnyHashable: Any]) -> Void)?
        
        public init(config: Configuration) {
            self.config = config
            self.onConfigurationSuccess = config.onConfigurationSuccess
            self.onStoreConnectionEvent = config.onStoreConnectionEvent
            self.onInvoiceEvent = config.onInvoiceEvent
            self.onCheckingStoreConnectionEvent = config.onCheckingStoreConnectionEvent
            setupNotificationObservers()
        }
        
        deinit {
          removeNotificationObservers()
        }
        
        public class SDKViewController: StorelinkViewController {
        }

        public func getBackgroundView() -> StorelinkCoreView {
            return StorelinkCoreView(refreshToken: config.refreshToken, logLevel: config.logLevel, viewType: .backgroundTask, _devApiLocation: config._devApiLocation)
        }
//        
//        public func getStoreConnectionsListView() -> StorelinkCoreView {
//            return StorelinkCoreView(refreshToken: config.refreshToken, logLevel: config.logLevel, viewType: .storeConnectionsList)
//        }
      
        public func getConnectUpdateStoreView(storeId: String, onComplete: (([AnyHashable: Any]) -> Void)? = nil) -> StorelinkCoreView {
            return StorelinkCoreView(refreshToken: config.refreshToken, logLevel: config.logLevel, viewType: .connectUpdateStore, functionParams: ["storeId": storeId], onComplete: onComplete, brandName: config.brandName, logoUrl: config.logoUrl, _devApiLocation: config._devApiLocation)
        }
      
        public func getTransferCartView(cartId: String, onComplete: (([AnyHashable: Any]) -> Void)? = nil) -> StorelinkCoreView {
            return StorelinkCoreView(refreshToken: config.refreshToken, logLevel: config.logLevel, viewType: .transferCart, functionParams: ["cartId": cartId], onComplete: onComplete, brandName: config.brandName, logoUrl: config.logoUrl, _devApiLocation: config._devApiLocation)
        }

        private func setupNotificationObservers() {
            NotificationCenter.default.addObserver(forName: self.notificationName, object: nil, queue: .main) { notification in
                if let params = notification.userInfo as? [String: Any],
                let functionName = params["functionName"] as? String {
                    switch functionName {
                    case "onConfigurationSuccess":
                        self.onConfigurationSuccess?(params)
                    case "onStoreConnectionEvent":
                        self.onStoreConnectionEvent?(params)
                    case "onInvoiceEvent":
                        self.onInvoiceEvent?(params)
                    case "onCheckingStoreConnectionEvent":
                        self.onCheckingStoreConnectionEvent?(params)
                    default:
                        break // or handle unknown functionName
                    }
                }
            }
        }
        
        private func removeNotificationObservers() {
            NotificationCenter.default.removeObserver(self, name: self.notificationName, object: nil)
        }

        public func sendDataToReactNative(data: [AnyHashable: Any]) {
            NotificationCenter.default.post(name: Notification.Name("CooklistDataFromNative"), object: nil, userInfo: data)
        }

        // public func sendDataToReactNativeAndWait(data: [AnyHashable: Any]) async throws -> [AnyHashable: Any]? {

        //     sendDataToReactNative(data: data)

        //     // Use Task for timeout handling; 10s limit
        //     return try await withCheckedThrowingContinuation { continuation in
        //         let timeoutTask = Task {
        //             try await Task.sleep(nanoseconds: UInt64(10_000_000_000))
        //             continuation.resume(throwing: SDKError.nativeCommunicationError)
        //         }

        //         NotificationCenter.default.addObserver(forName: self.notificationName, object: nil, queue: .main) { notification in
        //             timeoutTask.cancel()  // Cancel the timeout task if we receive a response

        //             if let params = notification.userInfo {
        //                 continuation.resume(returning: params)
        //             }
        //         }
        //     }
        // }

        // The method to open a UI based on the specified presentation method
        // public func open(presentUsing method: PresentationMethod) {
        //     let sdkUI = StorelinkViewController()
            
        //     switch method {
        //     case .presentModally(let parentVC):
        //         let navigationController = UINavigationController(rootViewController: sdkUI)
        //         parentVC.present(navigationController, animated: true)
                
        //     case .pushOnNavigationStack(let navigationController):
        //         navigationController.pushViewController(sdkUI, animated: true)
                
        //     case .embedInTab(let tabBarController, let position):
        //         if let index = position, index < tabBarController.viewControllers?.count ?? 0 {
        //             tabBarController.viewControllers?.insert(sdkUI, at: index)
        //         } else {
        //             tabBarController.viewControllers?.append(sdkUI)
        //         }
        //     }
        // }
    }
    
    // Create a handler based on configuration
    public static func create(_ config: Configuration) async -> Result<SDKHandler, SDKError> {
        do {
          let handler = SDKHandler(config: config)
          return .success(handler)
        } catch let error as SDKError {
            // Handle specific SDKError
            return .failure(error)
        } catch {
            // Handle other errors
            return .failure(.unknownError)
        }
    }
}

File: android/storelink/src/main/java/com/storelink/StorelinkCore.java
package com.storelink;

import android.content.Context;
import android.content.Intent;
import androidx.fragment.app.Fragment;

import java.util.Map;

public class StorelinkCore {

    // Configuration for the SDK
    public static class Configuration {
        private String refreshToken;
        private LogLevel logLevel;
        private OnEventListener onConfigurationSuccess;
        private OnEventListener onStoreConnectionEvent;
        private OnEventListener onInvoiceEvent;
        private OnEventListener onCheckingStoreConnectionEvent;
        private String brandName;
        private String logoUrl;
        private String devApiLocation;

        public Configuration(String refreshToken, LogLevel logLevel, OnEventListener onConfigurationSuccess, OnEventListener onStoreConnectionEvent, OnEventListener onInvoiceEvent, OnEventListener onCheckingStoreConnectionEvent, String brandName, String logoUrl, String devApiLocation) {
            this.refreshToken = refreshToken;
            this.logLevel = logLevel;
            this.onConfigurationSuccess = onConfigurationSuccess;
            this.onStoreConnectionEvent = onStoreConnectionEvent;
            this.onInvoiceEvent = onInvoiceEvent;
            this.onCheckingStoreConnectionEvent = onCheckingStoreConnectionEvent;
            this.brandName = brandName;
            this.logoUrl = logoUrl;
            this.devApiLocation = devApiLocation;
        }

        public String getRefreshToken() {
            return refreshToken;
        }

        public LogLevel getLogLevel() {
            return logLevel;
        }

        public OnEventListener getOnConfigurationSuccess() {
            return onConfigurationSuccess;
        }

        public OnEventListener getOnStoreConnectionEvent() {
            return onStoreConnectionEvent;
        }

        public OnEventListener getOnInvoiceEvent() {
            return onInvoiceEvent;
        }

        public OnEventListener getOnCheckingStoreConnectionEvent() {
            return onCheckingStoreConnectionEvent;
        }

        public String getBrandName() {
            return brandName;
        }

        public String getLogoUrl() {
            return logoUrl;
        }

        public String getDevApiLocation() {
            return devApiLocation;
        }
    }

    // Possible errors for the SDK
    public enum SDKError {
        INVALID_REFRESH_TOKEN,
        UNKNOWN_ERROR,
        NATIVE_COMMUNICATION_ERROR
    }

    // Define a handler that the SDK uses
    public static class SDKHandler {

        private Configuration config;

        public SDKHandler(Configuration config) {
            this.config = config;
            setupNotificationObservers();
        }

        private void setupNotificationObservers() {
            EventManager.getInstance().getEventLiveData().observeForever(event -> {
                if (event != null && event.containsKey("functionName")) {
                    String functionName = (String) event.get("functionName");
                    switch (functionName) {
                        case "onConfigurationSuccess":
                            if (config.getOnConfigurationSuccess() != null) {
                                config.getOnConfigurationSuccess().onEvent(event);
                            }
                            break;
                        case "onStoreConnectionEvent":
                            if (config.getOnStoreConnectionEvent() != null) {
                                config.getOnStoreConnectionEvent().onEvent(event);
                            }
                            break;
                        case "onInvoiceEvent":
                            if (config.getOnInvoiceEvent() != null) {
                                config.getOnInvoiceEvent().onEvent(event);
                            }
                            break;
                        case "onCheckingStoreConnectionEvent":
                            if (config.getOnCheckingStoreConnectionEvent() != null) {
                                config.getOnCheckingStoreConnectionEvent().onEvent(event);
                            }
                            break;
                        default:
                            break;
                    }
                }
            });
        }

        public Fragment getBackgroundView() {
            return StorelinkCoreViewFragment.newInstance(config.getRefreshToken(), config.getLogLevel(), ViewType.BACKGROUND_TASK, null);
        }

        public Fragment getConnectUpdateStoreView(String storeId, OnCompleteListener onComplete) {
            return StorelinkCoreViewFragment.newInstance(config.getRefreshToken(), config.getLogLevel(), ViewType.CONNECT_UPDATE_STORE, Map.of("storeId", storeId), onComplete);
        }

        public Fragment getTransferCartView(String cartId, OnCompleteListener onComplete) {
            return StorelinkCoreViewFragment.newInstance(config.getRefreshToken(), config.getLogLevel(), ViewType.TRANSFER_CART, Map.of("cartId", cartId), onComplete);
        }

        public void sendDataToReactNative(Map<String, Object> data) {
            EventManager.getInstance().postEvent(data);
        }
    }

    // Create a handler based on configuration
    public static SDKHandler create(Configuration config) {
        return new SDKHandler(config);
    }
}
